File: elixir/luhn/luhn.exs
defmodule Luhn do

  @doc """
  Checks if the given number is valid via the luhn formula
  """
  @spec valid?(String.t()) :: boolean
  def valid?(number) do
    number = clean(number)

    cond do
      String.length(number) < 2 -> false
      String.match?(number, ~r/[^0-9]/) -> false
      true -> check(number)
    end
  end

  def check(number) do
    sum =
      number
      |> String.graphemes
      |> Enum.map(&String.to_integer(&1))
      |> Kernel.++([0])
      |> Enum.reverse
      |> Enum.map_every(2, &(&1*2))
      |> Enum.map(fn n -> if n > 9, do: n-9, else: n end)
      |> Enum.sum

    case rem(sum, 10) do
      0 -> true
      _ -> false
    end
  end

  def clean(number) do
    number
    |> String.replace(" ", "")
  end
end

File: python/luhn/luhn.py
import re


class Luhn(object):
    def __init__(self, card_num):
        self.number = card_num

    def is_valid(self):
        digits = [int(n) for n in re.findall(r'\d', self.number)][::-1]

        if re.findall(r'[^0-9\s]', self.number):
            return False
        elif len(digits) < 2:
            return False

        check_digits = [n * 2 for n in digits[1::2]]
        check_digits = [n - 9 if n > 9 else n for n in check_digits]
        digits = [n for n in digits[::2]]

        return ((sum(digits) + sum(check_digits)) % 10) == 0
